File: Shortest Path in Undirected Graph.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    vector<int> shortestPath(vector<vector<int>>& adj, int src) {
        int n = adj.size();
        vector<int> res(n, -1);
        queue<int> q;

        res[src] = 0;
        q.push(src);

        while (!q.empty()) {
            int top = q.front();
            q.pop();

            for (int nb : adj[top]) {
                if (res[nb] == -1) {
                    res[nb] = res[top] + 1;
                    q.push(nb);
                }
            }
        }

        return res;
    }
};

File: Shortest Path in Undirected Graph.java

class Solution {
    // Function to find the shortest path from a source node to all other nodes
    public int[] shortestPath(ArrayList<ArrayList<Integer>> adj, int src) {
        // code here
        HashMap<Integer, ArrayList<Integer>> hm = new HashMap<>();
        int n = adj.size();
        for(int i = 0; i < n; i++){
            ArrayList<Integer> arr = new ArrayList<>();
            for(int j = 0; j < adj.get(i).size(); j++){
                arr.add(adj.get(i).get(j));
            }
            hm.put(i, arr);
        }
        
        Queue<Integer> q = new LinkedList<>();
        q.offer(src);
        int[] res = new int[n];
        for(int i = 0; i < n; i++) res[i] = -1;
        res[src] = 0;
        
        while(!q.isEmpty()){
            int top = q.poll();
            if(hm.get(top).size() != 0){
                for(int nb: hm.get(top)){
                    if(res[nb] == -1){
                        q.offer(nb);
                        res[nb] = res[top] + 1;
                    }
                }
            }
        }
        
        return res;
    }
}
